app/chat.hoon

```diff
@@ -0,0 +1,606 @@
+            |=  src=source:hall
+            ^-  circle:hall
+            cir.src
+        ::
+        =/  meslis/(list [circle:hall (list envelope:hall)])
+          %+  turn  circles
+          |=  cir=circle:hall
+          ^-  [circle:hall (list envelope:hall)]
+          [cir ~]
+        ::
+        =/  localpeers/(set @p)  
+          %-  silt  %+  turn  ~(tap by loc.pes.piz)
+          |=  [shp=@p stat=status:hall]
+          shp
+        ::
+        =/  peers/(map circle:hall (set @p))
+          %-  ~(rep by rem.pes.piz)
+          |=  [[cir=circle:hall grp=group:hall] acc=(map circle:hall (set @p))]
+          ^-  (map circle:hall (set @p))
+          =/  newset
+            %-  silt  %+  turn  ~(tap by grp)
+            |=  [shp=@p stat=status:hall]
+            shp
+          (~(put by acc) cir newset)
+        ::
+        :-
+          %+  turn  ~(tap in (~(del in (silt circles)) [our.bol %inbox]))
+            |=  cir=circle:hall
+            ^-  move
+            =/  wir/wire  /circle/[(scot %p our.bol)]/[nom.cir]/config/group
+            =/  pat/path  /circle/[nom.cir]/config/group
+            [ost.bol %peer wir [our.bol %hall] pat]
+          ::
+          %=  this
+              inbox.str.sta  loc.cos.piz
+              configs.str.sta  configs
+              messages.str.sta  (molt meslis)
+              peers.str.sta  (~(put by peers) [our.bol %inbox] localpeers)
+          ==
+    ::
+    ::  fill remote configs with message data
+    ::
+      =*  messages  messages.str.sta
+      =/  circle/circle:hall  [`@p`(slav %p &2:wir) &3:wir]
+      =/  localpeers/(set @p)  
+        %-  silt  %+  turn  ~(tap by loc.pes.piz)
+        |=  [shp=@p stat=status:hall]
+        shp
+      ::
+      =/  peers/(map circle:hall (set @p))
+        %-  ~(rep by rem.pes.piz)
+        |=  [[cir=circle:hall grp=group:hall] acc=(map circle:hall (set @p))]
+        ^-  (map circle:hall (set @p))
+        =/  newset
+          %-  silt  %+  turn  ~(tap by grp)
+          |=  [shp=@p stat=status:hall]
+          shp
+        (~(put by acc) cir newset)
+      :-  ~
+      %=  this
+        messages.str.sta  (~(put by messages) circle nes.piz)
+        peers.str.sta  (~(uni by peers.str.sta) (~(put by peers) circle localpeers))
+      ==
+  ==
+::
+::  +diff-hall-rumor: handle updates to hall state
+::
+++  diff-hall-rumor
+  |=  [wir=wire rum=rumor:hall]
+  ^-  (quip move _this)
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ::
+  ::  %circles
+      %circles
+    ?>  ?=(%circles -.rum)
+      =/  cis
+        ?:  add.rum
+          (~(put in circles.str.sta) cir.rum)
+        (~(del in circles.str.sta) cir.rum)
+      =/  str  %=  str.sta
+          circles  cis
+          peers
+            ?:  add.rum
+              (~(put by peers.str.sta) [our.bol cir.rum] ~)
+            (~(del by peers.str.sta) [our.bol cir.rum])
+      ==
```

Comment:
This `=/` and its inner `%=` makes confusing use of whitespace, please indent/backstep idiomatically.

Reply:
fixed